pkg/scheduler/testsuite_scheduler.go

```diff
@@ -106,85 +108,102 @@ func (s *Scheduler) runSteps(ctx context.Context, wg *sync.WaitGroup, testsuiteE
 	s.logger.Infow("Running steps", "test", testsuiteExecution.Name)
 
 	hasFailedSteps := false
-	cancelSteps := false
+	var cancelSteps atomic.Bool
+	cancelSteps.Store(false)
 	var batchStepResult *testkube.TestSuiteBatchStepExecutionResult
 
 	var abortionStatus *testkube.TestSuiteExecutionStatus
-	abortChan := make(chan *testkube.TestSuiteExecutionStatus)
 
-	go s.abortionCheck(ctx, testsuiteExecution, request.Timeout, abortChan)
+	go s.timeoutCheck(ctx, testsuiteExecution, request.Timeout)
 
-	for i := range testsuiteExecution.ExecuteStepResults {
-		batchStepResult = &testsuiteExecution.ExecuteStepResults[i]
-		select {
-		case abortionStatus = <-abortChan:
-			s.logger.Infow("Aborting test suite execution", "execution", testsuiteExecution.Id, "i", i)
+	err := bus.Internal.SubscribeTopic(bus.InternalSubscribeTopic, testsuiteExecution.Name, func(event testkube.Event) error {
+		s.logger.Infow("test suite abortion event in runSteps", "event", event)
+		if event.TestSuiteExecution != nil &&
+			event.TestSuiteExecution.Id == testsuiteExecution.Id &&
+			event.Type_ != nil &&
+			(*event.Type_ == testkube.END_TESTSUITE_ABORTED_EventType || *event.Type_ == testkube.END_TESTSUITE_TIMEOUT_EventType) {
+			s.logger.Infow("Aborting test suite execution", "execution", testsuiteExecution.Id)
 
-			cancelSteps = true
+			cancelSteps.Store(true)
 			for j := range batchStepResult.Execute {
 				if batchStepResult.Execute[j].Execution != nil && batchStepResult.Execute[j].Execution.ExecutionResult != nil {
 					batchStepResult.Execute[j].Execution.ExecutionResult.Abort()
 				}
 			}
 
 			testsuiteExecution.Status = testkube.TestSuiteExecutionStatusAborting
-		default:
-			s.logger.Debugw("Running batch step", "step", batchStepResult.Execute, "i", i)
 
-			if cancelSteps {
-				s.logger.Debugw("Aborting batch step", "step", batchStepResult.Execute, "i", i)
+			abortionStatus = testkube.TestSuiteExecutionStatusAborting
+			if event.Type_ == testkube.EventEndTestTimeout {
+				abortionStatus = testkube.TestSuiteExecutionStatusTimeout
+			}
+		}
+		s.logger.Infow("test suite abortion event in runSteps end", "event", event)
+		return nil
+	})
 
-				for j := range batchStepResult.Execute {
-					if batchStepResult.Execute[j].Execution != nil && batchStepResult.Execute[j].Execution.ExecutionResult != nil {
-						batchStepResult.Execute[j].Execution.ExecutionResult.Abort()
-					}
-				}
+	if err != nil {
+		s.logger.Errorw("error subscribing to event", "error", err)
+	}
 
-				continue
-			}
+	for i := range testsuiteExecution.ExecuteStepResults {
+		batchStepResult = &testsuiteExecution.ExecuteStepResults[i]
+		s.logger.Debugw("Running batch step", "step", batchStepResult.Execute, "i", i)
+
+		if cancelSteps.Load() {
```

Comment:
Does this code  depends from SubscribeTopic? It should be async, no?

Reply:
Do we wait as in code below/

Reply:
fixed, it uses channels to sync and handler has no logic in it